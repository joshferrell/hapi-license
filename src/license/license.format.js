import joi from 'joi';

export const computerFormat = {
    computerId: joi
        .string()
        .required()
        .description('unique identify for machine based off of computer UUID'),
    computerUsername: joi
        .string()
        .required()
        .description('username for computer that app is in use for')
        .example('Paul'),
    computerOS: joi
        .string()
        .valid(['Windows', 'MacOS', 'Linux'])
        .required()
        .description('operating system that machine is using')
        .example('MacOS'),
    computerName: joi
        .string()
        .required()
        .description('name that user has defined for their computer')
        .example('Paul\'s Macbook Pro')
};

export const licenseFormat = joi.object({
    id: joi.string().uuid(['uuidv4']).required()
        .description('license id')
        .example('1dfc0c66-bcc0-4c02-aa0a-f633766f2dc5'),
    accountId: joi.string().required()
        .description('account user id')
        .example('1234'),
    productId: joi.string().required()
        .description('product id')
        .example('1bccad9a-10e5-41bb-9032-5f3a0e47ce9f'),
    expiresAt: joi.date().min('now').required()
        .description('date that license expires')
        .example('2999-01-01'),
    ...computerFormat,
    license: joi.string().required()
        .description('user license text to be imported into application'),
    createdAt: joi.date().required()
        .description('time that license was created')
        .example('2017-01-01T00:00:00.000Z'),
    updatedAt: joi.date().required()
        .description('time that license was updated')
        .example('2017-01-01T00:00:00.000Z'),
    url: joi.string().uri().required()
        .description('url that license is accessible')
        .example('http://localhost:3000/license/1dfc0c66-bcc0-4c02-aa0a-f633766f2dc5')
});
