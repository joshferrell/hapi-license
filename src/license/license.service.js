import { oneLineTrim } from 'common-tags';
import moment from 'moment';
import { promisify } from 'util';
import jwt from 'jsonwebtoken';

const sign = promisify(jwt.sign);

export const mapUrl = license => Object.assign({}, license, {
    url: oneLineTrim`
        ${process.env.SERVER_SCHEME}://
        ${process.env.SERVER_URL}/license/${license.id}
    `
});

export const createLicenseIssuer = () => ({
    accountId,
    validUntil,
    computer,
    productId
}) => sign(
    {
        accountId,
        computer,
        productId,
        exp: Number.parseInt(moment(validUntil).format('x'), 10)
    },
    'shhh',
    { algorithm: 'HS256' }
);
