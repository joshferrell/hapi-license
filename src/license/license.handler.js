import { pipe, map } from 'ramda';
import boom from 'boom';
import { mapUrl, createLicenseIssuer } from './license.service';

export const createFetchLicenses = LicenseModel => (request, reply) => {
    const {
        email,
        app_metadata: {
            licenseCount,
            licenseTotal
        }
    } = request.auth.credentials;

    return LicenseModel
        .findAll({
            where: { email },
            raw: true
        })
        .then(pipe(
            map(mapUrl),
            licenses => ({
                meta: { licenseCount, licenseTotal }, licenses
            }),
            reply
        ))
        .catch((err) => {
            request.log('error', {
                err,
                msg: 'unable to fetch licenses'
            });
            return reply(boom.badImplementation());
        });
};

export const createNewLicense = LicenseModel => (request, reply) => {
    const issueLicense = createLicenseIssuer();
    const { productId, expiresAt, computer } = request.payload;
    const {
        email,
        app_metadata: {
            licenseCount,
            licenseTotal
        }
    } = request.auth.credentials;

    if (licenseCount >= licenseTotal) {
        return reply(boom.unauthorized('request exceeds available license amount'));
    }

    return issueLicense({ ...request.payload, email })
        .then(license => LicenseModel.create({
            productId, expiresAt, ...computer, license, email
        }))
        .then(_ => _.get({ plain: true }))
        .then(pipe(
            mapUrl,
            license => ({
                meta: {
                    licenseCount: licenseCount + 1,
                    licenseTotal
                },
                license
            }),
            reply
        ))
        .catch((err) => {
            request.log('error', {
                err,
                msg: 'unable to create license'
            });
            return reply(boom.badImplementation());
        });
};

export const createDeleteLicense = LicenseModel => (request, reply) => {
    const { id } = request.params;
    const { email } = request.auth.credentials;

    return LicenseModel
        .destroy({
            where: { id, email }
        })
        .then(rowsAffected => (
            rowsAffected ?
                reply({ success: true }) :
                reply(boom.badRequest('license does not exist'))
        ))
        .catch((err) => {
            request.log('error', {
                err,
                msg: 'unable to delete license'
            });
            return reply(boom.badImplementation());
        });
};

export const createFetchLicense = LicenseModel => (request, reply) => {
    const { id } = request.params;
    const {
        email,
        app_metadata: {
            licenseCount,
            licenseTotal
        }
    } = request.auth.credentials;

    return LicenseModel
        .find({
            where: { email, id },
            raw: true
        })
        .then(license => (
            license ?
                pipe(
                    mapUrl,
                    x => ({
                        meta: { licenseCount, licenseTotal },
                        license: x
                    }),
                    reply
                )(license) :
                reply(boom.notFound())
        ))
        .catch((err) => {
            request.log('error', {
                err,
                msg: 'fetch inidividual license'
            });
            return reply(boom.badImplementation());
        });
};

export const createUpdateLicense = LicenseModel => (request, reply) => {
    const { id } = request.params;
    const { expiresAt, productId } = request.payload;
    const {
        email,
        app_metadata: {
            licenseCount,
            licenseTotal
        }
    } = request.auth.credentials;
    const issueLicense = createLicenseIssuer();

    return LicenseModel
        .findOne({ where: { email, id }, raw: true })
        .then((model) => {
            if (!model) return Promise.reject(Error('license does not exist'));

            const {
                computerId,
                computerUsername,
                computerOS,
                computerName
            } = model;

            const license = issueLicense({
                email,
                expiresAt: expiresAt || model.expiresAt,
                productId: productId || model.productId,
                computer: {
                    computerId,
                    computerUsername,
                    computerOS,
                    computerName
                }
            });

            return model.update({
                expiresAt: expiresAt || model.expiresAt,
                productId: productId || model.productId,
                license
            });
        })
        .then(model => model.get({ plain: true }))
        .then(license => reply({
            meta: {
                licenseCount, licenseTotal
            },
            license
        }))
        .catch((err) => {
            switch (err.message) {
            case 'license does not exist':
                request.log('debug', `requested license ${id} does not exist for user ${email}`);
                return reply(boom.badRequest());
            default:
                request.log('error', {
                    err,
                    msg: 'fetch inidividual license'
                });
                return reply(boom.badImplementation());
            }
        });
};
