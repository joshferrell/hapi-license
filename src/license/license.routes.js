import joi from 'joi';
import { format } from '../utility';
import { licenseFormat, computerFormat } from './license.format';
import {
    createFetchLicenses,
    createNewLicense,
    createDeleteLicense,
    createFetchLicense,
    createUpdateLicense
} from './license.handler';

const createLicenseRoutes = (LicenseModel) => {
    const fetchLicenses = createFetchLicenses(LicenseModel);
    const newLicense = createNewLicense(LicenseModel);
    const deleteLicense = createDeleteLicense(LicenseModel);
    const fetchLicense = createFetchLicense(LicenseModel);
    const updateLicense = createUpdateLicense(LicenseModel);

    return [
        {
            method: 'POST',
            path: '/license',
            handler: newLicense,
            config: {
                auth: {
                    scope: 'write:license'
                },
                validate: {
                    payload: {
                        productId: joi.string().required()
                            .description('product id associated with license')
                            .example('1bccad9a-10e5-41bb-9032-5f3a0e47ce9f'),
                        validUntil: joi.date().min('now').required()
                            .description('date that the license expires')
                            .example('2999-01-01'),
                        computer: computerFormat
                    }
                },
                tags: ['api', 'License Management'],
                description: 'Create a new license for a user',
                notes: [
                    'Returns a license for a user and creates a new license in the database'
                ],
                plugins: {
                    'hapi-swagger': {
                        responses: {
                            200: {
                                description: 'Creation successful',
                                schema: licenseFormat
                            },
                            ...format.badRequest,
                            ...format.unauthorized,
                            ...format.internalError,
                            ...format.notImplemented
                        }
                    }
                }
            }
        },
        {
            method: 'GET',
            path: '/license',
            handler: fetchLicenses,
            config: {
                auth: {
                    scope: 'read:license'
                },
                tags: ['api', 'License Management'],
                description: 'Get all active licenses for a user',
                notes: [
                    'Returns all licenses for a user as well as how many licenses the user has available'
                ],
                plugins: {
                    'hapi-swagger': {
                        responses: {
                            200: {
                                description: 'Fetch all licenses for user',
                                schema: joi.array().items(licenseFormat).required()
                            },
                            ...format.badRequest,
                            ...format.unauthorized,
                            ...format.internalError,
                            ...format.notImplemented
                        }
                    }
                }
            }
        },
        {
            method: 'GET',
            path: '/license/{id}',
            handler: fetchLicense,
            config: {
                auth: {
                    scope: 'read:license'
                },
                validate: {
                    params: {
                        id: joi.string().uuid(['uuidv4']).required()
                            .description('license id')
                            .example('1e589ff5-b19a-4642-a5b2-8db48c0aa1c4')
                    }
                },
                tags: ['api', 'License Management'],
                description: 'Get a license by the uuid',
                notes: [
                    'Returns all of the information about a single license'
                ],
                plugins: {
                    'hapi-swagger': {
                        responses: {
                            200: {
                                description: 'Fetch license',
                                schema: licenseFormat
                            },
                            ...format.badRequest,
                            ...format.unauthorized,
                            ...format.internalError,
                            ...format.notImplemented
                        }
                    }
                }
            }
        },
        {
            method: 'DELETE',
            path: '/license/{id}',
            handler: deleteLicense,
            config: {
                auth: {
                    scope: 'write:license'
                },
                validate: {
                    params: {
                        id: joi.string().uuid(['uuidv4']).required()
                            .description('license id')
                            .example('1e589ff5-b19a-4642-a5b2-8db48c0aa1c4')
                    }
                },
                tags: ['api', 'License Management'],
                description: 'Delete a license by uuid',
                notes: [
                    'Deletes a license for a user allowing the user to allocate an additional license'
                ],
                plugins: {
                    'hapi-swagger': {
                        responses: {
                            ...format.success,
                            ...format.badRequest,
                            ...format.unauthorized,
                            ...format.internalError,
                            ...format.notImplemented
                        }
                    }
                }
            }
        },
        {
            method: 'PATCH',
            path: '/license/{id}',
            handler: updateLicense,
            config: {
                auth: {
                    scope: 'write:license'
                },
                validate: {
                    params: {
                        id: joi.string().uuid(['uuidv4']).required()
                            .description('license id')
                            .example('1e589ff5-b19a-4642-a5b2-8db48c0aa1c4')
                    },
                    payload: joi.object({
                        validUntil: joi.date().min('now')
                            .description('date that license expires')
                            .example('2099-01-01'),
                        productId: joi.string()
                            .description('product that user has purchased used')
                            .example('1234')
                    }).or('validUntil', 'productId').required()
                },
                tags: ['api', 'License Management'],
                description: 'Update license expiration',
                notes: [
                    'valid until date set in case that license has been updated for expire time',
                    'product id set in case the user has updated their product of choice'
                ],
                plugins: {
                    'hapi-swagger': {
                        responses: {
                            200: {
                                description: 'Creation successful',
                                schema: licenseFormat
                            },
                            ...format.badRequest,
                            ...format.unauthorized,
                            ...format.internalError,
                            ...format.notImplemented
                        }
                    }
                }
            }
        }
    ];
};

export default createLicenseRoutes;
