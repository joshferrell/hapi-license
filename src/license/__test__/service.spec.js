import jwt from 'jsonwebtoken';
import { promisify } from 'util';
import { mapUrl, createLicenseIssuer } from '../license.service';

const verify = promisify(jwt.verify);

describe('license services', () => {
    beforeEach(() => {
        process.env.SERVER_SCHEME = 'https';
        process.env.SERVER_URL = 'localhost:3000';
    });

    describe('license map url', () => {
        it('should return the url for a particular license object', () => {
            const license = {
                test: 'Hi I\'m Paul!',
                id: '1234'
            };

            expect(mapUrl(license)).toEqual({
                test: 'Hi I\'m Paul!',
                id: '1234',
                url: 'https://localhost:3000/license/1234'
            });
        });
    });

    describe('issue license', async () => {
        const issueLicense = createLicenseIssuer();
        const data = {
            accountId: '1234',
            expiresAt: '2099-01-01',
            computer: 'Paul\'s Macbook',
            productId: 'awesome product!'
        };

        const license = await issueLicense(data);
        const decoded = await verify(license, 'shhh');
        expect(decoded).toMatchSnapshot();
    });
});
