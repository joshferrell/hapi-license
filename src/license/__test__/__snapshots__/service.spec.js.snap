// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`license services license map url should return the url for a particular license object 1`] = `
Object {
  "accountId": "1234",
  "computer": "Paul's Macbook",
  "exp": 4070937600000,
  "iat": 1508729273,
  "productId": "awesome product!",
}
`;
